doc/designs/app-passwords.md

```diff
@@ -0,0 +1,312 @@
+
+## Feature management
+
+The login process will remain the same, only instead of entering the primary
+password, the user will enter an app password. And as already stated, the user
+then will not have access to the account management.
+
+### UI
+
+TBD
+
+### CLI
+
+| Command | Arguments | Options |
+| ------- | --------- | ------- |
+| apppw-add | username (user's uid), displayname, appname |  |
+| apppw-list | username (user's uid) |  |
+| apppw-del | username (user's uid), uuid |  |
+
+## Test Plan
+
+1. All existing tests must pass. This also ensures that the user can have only
+one password.
+2. Test app passwords:
+   1. Add one app password using WebUI and one using CLI.
+   2. Check that using these app passwords, it is possible to authenticate to
+      all services and systems, but not to the FreeIPA server itself for account
+      management.
+   3. Log in using the primary password.
+   4. Check whether all user's passwords are listed in the WebUI or by using an
+      appropriate command in CLI.
+   5. Delete both app passwords - the one created using WebUI by using CLI, and
+      vice versa.
+   6. Repeat steps 2.1 - 2.6 with using the both the deleted old and new app
+      passwords in step 2.2. Authentication using the deleted old app passwords
+      must fail, and authentication using the new app passwords must succeed.
+   7. Chceck that the user now only has the primary password.
+3. Test interactions with SSSD using the new functionality.
+   * Repeat all procedures from the point 2, but when testing logins to apps,
+     perform them on a client using SSSD configured against the testing FreeIPA
+     server.
+
+## Possible future enhancements
+
+### Support for use of the aforementioned functionality with Kerberos
+
+Support for use of the aforementioned functionality with Kerberos is necessary
+for HBAC to be applicable to users logged in using app passwords.
+
+Subaccounts will be introduced with principals in the form 
+```_user_/_appname_@_REALM_```
+(to be consistent with other things in FreeIPA) - a username will remain the
+same for subaccounts, only the principal will contain the app name. The code of
+FreeIPA must be modified to take into account the newly changed assumptions.
+Until now, these were:
+* two-component principals with a host as the first component are always host
+  principals (`_host_/_fqdn_`)
+* all other two-component principals are always service principals
+  (`_service_/_fqdn_`)
+* single-component principals are always users (user)
+Now, they will be:
+* two-component principals with an app name as the first component are users
+  logged in with an app password (`_user_/_appname_`)
+* two-component principals with a host as the first component are host
+  principals (`_host_/_fqdn_`)
+* all other two-component principals are service principals (`_service_/_fqdn_`)
+* single-component principals are users logged in with their primary password
+  (`user`)
```

Comment:
I think some examples of real principles would clarify what you're trying to say here.

In order to do a system login then something needs to resolve the appname LDAP entry into a POSIX account.

Reply:
OK, I will clarify it. BTW, I just realized that I may have misunderstood what a host principal is. Its first component is always the string 'host'?